### PR TITLE
DQC Rule 13 Updates

### PR DESCRIPTION
**Pull request by andrewperkins-wf**
# ULTIMATE PROBLEM
* DQC 13 needed updates to its pre-requisite configurations
* Needed to update rules that were updated to version 2.0
* Fixed some flake 8 errors

# HOW IT WAS FIXED
* Updated the version numbers in the rule files
* Updated the DQC 13 precondition configurations
* Fixed some Flake 8 failures

# TESTING SUGGESTIONS
* Test that DQC 13 is fixed

# AREAS AFFECTED
DQC 13, plus other rule definitions

# UNIT TESTS
None since this was a configuration change

### CODE REVIEW

File: dqc_us_rules/dqc_us_0001.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 from collections import defaultdict
 
 _CODE_NAME = 'DQC.US.0001'
-_RULE_VERSION = '1.1'
+_RULE_VERSION = '2.0'
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0001.py]: Per the semantic versioning thing in the expanded documentation, should this now be `2.0.0`?

File: dqc_us_rules/dqc_us_0013.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 
 
 _CODE_NAME = 'DQC.US.0013'
-_RULE_VERSION = '1.1'
+_RULE_VERSION = '2.0'
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0013.py]: See above comment

File: dqc_us_rules/dqc_us_0013.py [hunk 3]
```diff
@@ -23,34 +23,42 @@
     'dqc_15_exclusion_rules.csv'
 )
 
-_PRECONDITION_ELEMENTS = collections.OrderedDict([
+# This is setup as a list of tuples.  The first item is the concept to check
+# for existence of, the second is a list of concept(s) to sum and check if
+# they are greater than zero.
+_pre_cfg = [
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0013.py]: SG: Module level constants should be uppercase

File: dqc_us_rules/dqc_us_0018.py [hunk 4]
```diff
@@ -73,7 +73,7 @@ def _load_cache(val):
                 val.usgaapDeprecations = json.load(file)
                 file.close()
 
-            except FileNotFoundError:
+            except FileNotFoundError:  # noqa
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0018.py]: This weirds me out about flake8, to be honest, I wonder why it isn't seeing the builtin?

**Comment by chrislococo-wf** [comment]
@davidtauriello Do you mean the value in the actual markdown/CSV file names?

**Comment by davidtauriello** [comment]
@chrislococo-wf - no; the _rule_version = 1.0.0 or 2.0.0 from the python files should be copied to the corresponding row of the /docs/readme.md

No worries if it doesn't happen today - should be done for release though.

**Comment by chrislococo-wf** [comment]
@davidtauriello As in this line here https://github.com/DataQualityCommittee/dqc_us_rules/blame/master/docs/DQC_US_0001/DQC_0001.md#L62 ?

That is just sample output, would that need to be kept in lockstep with the current released version?

**Comment by davidtauriello** [comment]
@chrislococo-wf not the specific documentation, just the right column on this index: https://github.com/DataQualityCommittee/dqc_us_rules/blob/next_q1_16/docs/README.md - this is so users can check/confirm messages from results to ensure they're testing with the latest rule version. Lockstep = at each public exposure or approval release (yes).

**Comment by jasonleinbach-wf** [comment]
+1

**Comment by mindeerickson-wf** [comment]
+1

File: tests/unit_tests/test_dqc_us_0013.py [hunk 5]
```diff
@@ -114,11 +115,11 @@ def setUp(self):
             context=mock_context
         )
 
+
```
Comment (jasonleinbach-wf) [inline tests/unit_tests/test_dqc_us_0013.py]: this failed flake8

File: dqc_us_rules/dqc_us_0013.py [hunk 6]
```diff
@@ -100,28 +108,30 @@ def filter_negative_number_with_dependence_facts(val, blacklist_concepts):
...
-                continue  # cannot be black
-            if fact.qname.localName in blacklist_concepts:
-                bad_blacklist.append(fact)
-
+    else:
+        facts_to_check = []
+    # identify facts which should be reported as included in the list
+    for fact in facts_to_check:
+        if neg_num.check_rules(fact, blacklist_exclusion_rules):
+
```
Comment (jasonleinbach-wf) [inline dqc_us_rules/dqc_us_0013.py]: Unnecessary whitespace.

File: dqc_us_rules/dqc_us_0013.py [hunk 7]
```diff
@@ -133,24 +143,37 @@ def dqc_13_precondition_check(val):
 
     :param val: val whose modelXbrl provides the facts to check
     :type val: :class:'~arelle.ModelXbrl.ModelXbrl'
-    :return: True or False depending on the precondition check
-    :rtype: bool
+    :return: Returns the list of context(s) of the precondition fact(s)
+    :rtype: :class:'~arelle.ModelInstanceObject.ModelContext'
```
Comment (jasonleinbach-wf) [inline dqc_us_rules/dqc_us_0013.py]: This is now a list[class]

File: dqc_us_rules/dqc_us_0013.py [hunk 8]
```diff
@@ -133,24 +143,37 @@ def dqc_13_precondition_check(val):
...
-        if value > 0:
-            return True
-
-    return False
+    check_contexts = set()
+    seen_contexts = set()
+    pre_cache = {
+        x: {
+            y: [] for y in pre_checks
+            } for x, pre_checks in _PRECONDITION_ELEMENTS.items()
```
Comment (jasonleinbach-wf) [inline dqc_us_rules/dqc_us_0013.py]: This is pretty.

**Comment by jasonleinbach-wf** [comment]
Pass complete

File: dqc_us_rules/dqc_us_0013.py (hunk 8 above)
Comment (mindeerickson-wf) [inline dqc_us_rules/dqc_us_0013.py]: Thanks?!?...moving forward then

Comment (brianmyers-wf) [inline dqc_us_rules/dqc_us_0013.py]: looks like the closing } is one space to far indented...

File: dqc_us_rules/dqc_us_0013.py [hunk 9]
```diff
@@ -133,24 +143,37 @@ def dqc_13_precondition_check(val):
...
+        }
```
Comment (brianmyers-wf) [inline dqc_us_rules/dqc_us_0013.py]: looks like this closing } is off it's indentation

File: dqc_us_rules/dqc_us_0013.py (hunk 8 above)
Comment (mindeerickson-wf) [inline dqc_us_rules/dqc_us_0013.py]: Good call, Brian!...I'll fix it.  Thanks!

File: dqc_us_rules/dqc_us_0013.py (hunk 9 above)
Comment (mindeerickson-wf) [inline dqc_us_rules/dqc_us_0013.py]: Another good call:-)

**Comment by jasonleinbach-wf** [comment]
+1 pending travis

**Comment by brianmyers-wf** [comment]
+1

**Comment by jasonleinbach-wf** [comment]
+10 verified the docs attached to the ticket.

**Comment by jasonleinbach-wf** [comment]
Reviewed, verified.  Merging.
